File: explorer/darwin/ios/lynx_explorer/LynxExplorer/routing/LaunchDescriptorParser.swift
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

import Foundation

struct LaunchDescriptorParser {
  private let recorderURLPrefix: String
  private let maximumWrapperDepth: Int
  private let sparklingSchemeResolver: SparklingSchemeResolving

  init(
    recorderURLPrefix: String,
    maximumWrapperDepth: Int = 2,
    sparklingSchemeResolver: SparklingSchemeResolving = SparklingSchemeAdapter()
  ) {
    self.recorderURLPrefix = recorderURLPrefix
    self.maximumWrapperDepth = max(0, maximumWrapperDepth)
    self.sparklingSchemeResolver = sparklingSchemeResolver
  }

  func parse(
    _ input: String,
    requestedContainer: RequestedContainer,
    source: RouteSource
  ) throws -> LaunchDescriptor {
    try parse(
      input,
      originalInput: input,
      requestedContainer: requestedContainer,
      source: source,
      wrapperDepth: 0
    )
  }

  private func parse(
    _ input: String,
    originalInput: String,
    requestedContainer: RequestedContainer,
    source: RouteSource,
    wrapperDepth: Int
  ) throws -> LaunchDescriptor {
    let route = input.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !route.isEmpty else {
      throw RouteError.emptyInput
    }
    guard Self.hasValidPercentEncoding(route) else {
      throw RouteError.invalidPercentEncoding(route)
    }
    guard let components = URLComponents(string: route),
      let scheme = components.scheme?.lowercased(),
      !scheme.isEmpty
    else {
      throw RouteError.malformedURL(route)
    }

    if scheme == "lynx" {
      return try unwrap(
        route,
        components: components,
        originalInput: originalInput,
        requestedContainer: requestedContainer,
        source: source,
        wrapperDepth: wrapperDepth
      )
    }

    if !recorderURLPrefix.isEmpty && route.hasPrefix(recorderURLPrefix) {
      guard requestedContainer != .sparkling else {
        throw RouteError.recorderUnsupportedInSparkling
      }
      return try makeDescriptor(
        originalInput: originalInput,
        resource: .recorder(route),
        requestedContainer: requestedContainer,
        resolvedContainer: .legacy,
        source: source,
        canonicalSparklingScheme: nil,
        queryItems: Self.queryItems(from: components)
      )
    }

    switch scheme {
    case "http", "https":
      return try parseRemote(
        route,
        components: components,
        originalInput: originalInput,
        requestedContainer: requestedContainer,
        source: source
      )
    case "file":
      return try parseLocal(
        route,
        components: components,
        originalInput: originalInput,
        requestedContainer: requestedContainer,
        source: source
      )
    case "hybrid":
      return try parseCanonicalSparkling(
        route,
        originalInput: originalInput,
        requestedContainer: requestedContainer,
        source: source
      )
    default:
      throw RouteError.unsupportedScheme(scheme)
    }
  }

  private func unwrap(
    _ route: String,
    components: URLComponents,
    originalInput: String,
    requestedContainer: RequestedContainer,
    source: RouteSource,
    wrapperDepth: Int
  ) throws -> LaunchDescriptor {
    guard components.host?.lowercased() == "open" else {
      throw RouteError.unsupportedScheme(components.scheme?.lowercased() ?? "lynx")
    }
    let target: String
    if let legacyTarget = Self.unescapedLegacyWrapperTarget(from: components) {
      target = legacyTarget
    } else {
      let targets = Self.queryItems(from: components).filter { $0.name == "url" }
      guard targets.count <= 1 else {
        throw RouteError.ambiguousTarget("url")
      }
      guard let encodedTarget = targets.first?.value, !encodedTarget.isEmpty else {
        throw RouteError.missingTarget("url")
      }
      target = encodedTarget
    }
    guard wrapperDepth < maximumWrapperDepth else {
      throw RouteError.wrapperDepthExceeded(maximumWrapperDepth)
    }

    return try parse(
      target,
      originalInput: originalInput,
      requestedContainer: requestedContainer,
      source: source,
      wrapperDepth: wrapperDepth + 1
    )
  }

  private func parseRemote(
    _ route: String,
    components: URLComponents,
    originalInput: String,
    requestedContainer: RequestedContainer,
    source: RouteSource
  ) throws -> LaunchDescriptor {
    guard let host = components.host, !host.isEmpty, let url = URL(string: route) else {
      throw RouteError.malformedURL(route)
    }
    return try makeDescriptor(
      originalInput: originalInput,
      resource: .remote(url),
      requestedContainer: requestedContainer,
      resolvedContainer: requestedContainer == .sparkling ? .sparkling : .legacy,
      source: source,
      canonicalSparklingScheme: nil,
      queryItems: Self.queryItems(from: components)
    )
  }

  private func parseLocal(
    _ route: String,
    components: URLComponents,
    originalInput: String,
    requestedContainer: RequestedContainer,
    source: RouteSource
  ) throws -> LaunchDescriptor {
    guard components.host?.lowercased() == "lynx",
      let nestedRoute = components.percentEncodedQuery,
      !nestedRoute.isEmpty,
      Self.hasValidPercentEncoding(nestedRoute),
      let nestedComponents = URLComponents(string: nestedRoute),
      nestedComponents.scheme?.lowercased() == "local",
      let nestedHost = nestedComponents.host,
      !nestedHost.isEmpty,
      let decodedPath = nestedComponents.percentEncodedPath.removingPercentEncoding
    else {
      throw RouteError.malformedURL(route)
    }

    return try makeDescriptor(
      originalInput: originalInput,
      resource: .localBundle(nestedHost + decodedPath),
      requestedContainer: requestedContainer,
      resolvedContainer: requestedContainer == .sparkling ? .sparkling : .legacy,
      source: source,
      canonicalSparklingScheme: nil,
      queryItems: Self.queryItems(from: nestedComponents)
    )
  }

  private func parseCanonicalSparkling(
    _ route: String,
    originalInput: String,
    requestedContainer: RequestedContainer,
    source: RouteSource
  ) throws -> LaunchDescriptor {
    let resolution = try sparklingSchemeResolver.resolve(route)

    return try makeDescriptor(
      originalInput: originalInput,
      resource: resolution.resource,
      requestedContainer: requestedContainer,
      resolvedContainer: .sparkling,
      source: source,
      canonicalSparklingScheme: resolution.canonicalScheme,
      queryItems: resolution.outerQueryItems,
      normalizedExtras: resolution.normalizedExtras
    )
  }

  private func makeDescriptor(
    originalInput: String,
    resource: LaunchResource,
    requestedContainer: RequestedContainer,
    resolvedContainer: ResolvedContainer,
    source: RouteSource,
    canonicalSparklingScheme: String?,
    queryItems: [LaunchQueryItem],
    normalizedExtras: [String: String] = [:]
  ) throws -> LaunchDescriptor {
    let normalizedItems = normalizedExtras.sorted { $0.key < $1.key }.map {
      LaunchQueryItem(name: $0.key, value: $0.value)
    }
    let semanticItems = normalizedItems + queryItems
    let animated = Self.booleanValue(named: "animated", in: semanticItems) ?? true
    let fullScreen = Self.booleanValue(named: "fullscreen", in: semanticItems) ?? false
    let explicitlyHidden =
      Self.booleanValue(named: "hide_nav_bar", in: semanticItems)
      ?? Self.booleanValue(named: "hidden_nav", in: semanticItems)
      ?? false
    let statusBarHidden =
      Self.booleanValue(named: "hide_status_bar", in: semanticItems) ?? false
    let transparentStatusBar =
      Self.booleanValue(named: "trans_status_bar", in: semanticItems) ?? false
    let viewport = Self.viewport(from: semanticItems)
    let orientation = Self.orientation(from: semanticItems)

    var extras = normalizedExtras
    var pageGlobalProps: [String: AnyHashable] = [:]
    for (name, value) in normalizedExtras {
      pageGlobalProps[Self.globalPropName(for: name)] = value
    }
    for item in queryItems {
      guard let value = item.value else {
        continue
      }
      extras[item.name] = value
      pageGlobalProps[Self.globalPropName(for: item.name)] = value
    }

    return LaunchDescriptor(
      originalInput: originalInput,
      resource: resource,
      initialData: nil,
      commonGlobalProps: [:],
      pageGlobalProps: pageGlobalProps,
      pageName: Self.stringValue(named: "initial_page", in: semanticItems),
      viewport: viewport,
      appearance: AppearanceOptions(
        backgroundColor: Self.stringValue(named: "container_bg_color", in: semanticItems),
        transparentStatusBar: fullScreen || transparentStatusBar,
        statusBarHidden: fullScreen || statusBarHidden
      ),
      navigation: NavigationOptions(
        navigationHidden: fullScreen || explicitlyHidden,
        fullScreen: fullScreen,
        title: Self.stringValue(named: "title", in: semanticItems),
        titleColor: Self.stringValue(named: "title_color", in: semanticItems),
        barColor: Self.stringValue(named: "nav_bar_color", in: semanticItems)
          ?? Self.stringValue(named: "bar_color", in: semanticItems),
        backButtonStyle: Self.stringValue(named: "back_button_style", in: semanticItems),
        orientation: orientation
      ),
      presentation: PresentationOptions(animated: animated),
      cachePolicy: .useProtocolCachePolicy,
      queryItems: queryItems,
      extras: extras.mapValues(AnyHashable.init),
      requestedContainer: requestedContainer,
      container: resolvedContainer,
      source: source,
      canonicalSparklingScheme: canonicalSparklingScheme
    )
  }

  private static func queryItems(from components: URLComponents) -> [LaunchQueryItem] {
    (components.queryItems ?? []).map { LaunchQueryItem(name: $0.name, value: $0.value) }
  }

  /// The historical UIApplication entry treated everything after an
  /// unescaped `url=https://...&...` as the nested URL. Preserve that form
  /// while encoded wrappers continue through URLComponents, including its
  /// duplicate-target validation.
  private static func unescapedLegacyWrapperTarget(
    from components: URLComponents
  ) -> String? {
    guard let query = components.percentEncodedQuery else {
      return nil
    }

    var componentStart = query.startIndex
    while componentStart < query.endIndex {
      let componentEnd = query[componentStart...].firstIndex(of: "&") ?? query.endIndex
      let component = query[componentStart..<componentEnd]
      if component.hasPrefix("url=") {
        let valueStart = query.index(componentStart, offsetBy: 4)
        let rawTail = query[valueStart...]
        guard let separator = rawTail.firstIndex(of: "&") else {
          return nil
        }
        let firstValue = rawTail[..<separator]
        return firstValue.contains("://") ? String(rawTail) : nil
      }
      guard componentEnd < query.endIndex else {
        return nil
      }
      componentStart = query.index(after: componentEnd)
    }
    return nil
  }

  private static func stringValue(
    named name: String,
    in queryItems: [LaunchQueryItem]
  ) -> String? {
    queryItems.last(where: { $0.name == name && $0.value != nil })?.value
  }

  private static func booleanValue(
    named name: String,
    in queryItems: [LaunchQueryItem]
  ) -> Bool? {
    guard let rawValue = stringValue(named: name, in: queryItems) else {
      return nil
    }
    return (rawValue as NSString).boolValue
  }

  private static func viewport(from queryItems: [LaunchQueryItem]) -> ViewportOptions? {
    guard let widthValue = stringValue(named: "width", in: queryItems),
      let heightValue = stringValue(named: "height", in: queryItems),
      // The Legacy shell has always interpreted physical pixels with
      // NSString.intValue. Model that contract here so an explicit
      // Sparkling launch does not reinterpret inputs such as `1e3` as 1000
      // while the same Legacy launch renders at one physical pixel.
      case let width = (widthValue as NSString).intValue, width > 0,
      case let height = (heightValue as NSString).intValue, height > 0
    else {
      return nil
    }
    return ViewportOptions(
      widthInPixels: Double(width),
      heightInPixels: Double(height))
  }

  private static func orientation(from queryItems: [LaunchQueryItem]) -> LaunchOrientation? {
    guard let value = stringValue(named: "orientation", in: queryItems) else {
      return nil
    }
    // Preserve the Legacy shell's exact, lower-case-only contract.
    switch value {
    case "portrait":
      return .portrait
    case "landscape":
      return .landscape
    default:
      return nil
    }
  }

  private static func globalPropName(for queryName: String) -> String {
    let trimmedName = queryName.drop(while: { $0 == "_" })
    guard !trimmedName.isEmpty else {
      return queryName
    }

    let parts = trimmedName.split(separator: "_", omittingEmptySubsequences: true)
    guard let first = parts.first else {
      return queryName
    }
    return parts.dropFirst().reduce(String(first)) { result, part in
      guard let firstCharacter = part.first else {
        return result
      }
      return result + String(firstCharacter).uppercased() + part.dropFirst()
    }
  }

  private static func hasValidPercentEncoding(_ input: String) -> Bool {
    let bytes = Array(input.utf8)
    var index = 0
    while index < bytes.count {
      if bytes[index] == Character("%").asciiValue {
        guard index + 2 < bytes.count,
          isHexDigit(bytes[index + 1]),
          isHexDigit(bytes[index + 2])
        else {
          return false
        }
        index += 3
      } else {
        index += 1
      }
    }
    // Syntactically valid escapes can still encode invalid UTF-8 (for
    // example `%FF`). Foundation would silently turn those query values into
    // nil, which is not a lossless semantic descriptor.
    return input.removingPercentEncoding != nil
  }

  private static func isHexDigit(_ byte: UInt8) -> Bool {
    (byte >= Character("0").asciiValue! && byte <= Character("9").asciiValue!)
      || (byte >= Character("A").asciiValue! && byte <= Character("F").asciiValue!)
      || (byte >= Character("a").asciiValue! && byte <= Character("f").asciiValue!)
  }
}

File: explorer/android/lynx_explorer/src/main/java/com/lynx/explorer/routing/LaunchDescriptorParser.kt
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
package com.lynx.explorer.routing

import android.net.Uri
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

class LaunchDescriptorParser @JvmOverloads constructor(
  private val recorderPrefix: String = "sslocal://arkview",
  private val maximumWrapperDepth: Int = 2
) {
  fun parse(input: String, requestedRuntime: RequestedRuntime, source: RouteSource): LaunchDescriptor =
    parseInternal(input, input, requestedRuntime, source, 0)

  private fun parseInternal(input: String, original: String, requested: RequestedRuntime, source: RouteSource, depth: Int): LaunchDescriptor {
    val route = input.trim()
    if (route.isEmpty()) fail("empty_input", "Enter a Lynx bundle URL or a Sparkling hybrid scheme.")
    validatePercentEncoding(route)
    if ('%' in route) decode(route) // Validate percent-encoded bytes as UTF-8 before Uri normalizes them.
    val uri = try { Uri.parse(route) } catch (_: Exception) { fail("malformed_url", "Invalid absolute URL: $route") }
    val scheme = uri.scheme?.lowercase() ?: fail("malformed_url", "Invalid absolute URL: $route")
    if (scheme == "lynx") {
      if (!uri.host.equals("open", true)) fail("unsupported_scheme", "Unsupported Lynx route.")
      if (depth >= maximumWrapperDepth) fail("wrapper_depth_exceeded", "Too many nested lynx://open wrappers.")
      val target = wrapperTarget(uri)
      return parseInternal(target, original, requested, source, depth + 1)
    }
    if (recorderPrefix.isNotEmpty() && route.startsWith(recorderPrefix)) {
      if (requested == RequestedRuntime.SPARKLING) fail("recorder_unsupported_in_sparkling", "Recorder routes can only be opened with Lynx.")
      return descriptor(original, LaunchResource.Recorder(route), requested, ResolvedRuntime.LYNX, source, null, queryItems(uri))
    }
    return when (scheme) {
      "http", "https" -> {
        if (uri.host.isNullOrEmpty()) fail("malformed_url", "Remote bundle has no host.")
        descriptor(original, LaunchResource.RemoteBundle(route), requested, requested.resolveRaw(), source, null, queryItems(uri))
      }
      "assets" -> {
        val path = (uri.host.orEmpty() + uri.path.orEmpty()).trimStart('/')
        if (path.isEmpty()) fail("missing_target", "The route is missing its bundle target.")
        descriptor(original, LaunchResource.LocalBundle(path), requested, requested.resolveRaw(), source, null, queryItems(uri))
      }
      "file" -> parseFile(route, uri, original, requested, source)
      "hybrid" -> parseCanonical(route, uri, original, requested, source)
      else -> fail("unsupported_scheme", "The URL scheme '$scheme' is not supported by Lynx Explorer.")
    }
  }

  private fun parseFile(route: String, uri: Uri, original: String, requested: RequestedRuntime, source: RouteSource): LaunchDescriptor {
    if (!uri.host.equals("lynx", true)) fail("malformed_url", "Invalid local Lynx URL: $route")
    val nested = uri.encodedQuery ?: fail("missing_target", "The route is missing its local target.")
    validatePercentEncoding(nested)
    val local = Uri.parse(normalizeLegacyLocalQuery(nested))
    if (!local.scheme.equals("local", true)) fail("malformed_url", "Invalid local Lynx URL: $route")
    val path = (local.host.orEmpty() + local.path.orEmpty()).trimStart('/')
    if (path.isEmpty()) fail("missing_target", "The route is missing its local target.")
    return descriptor(original, LaunchResource.LocalBundle(path), requested, requested.resolveRaw(), source, null, queryItems(local))
  }

  private fun normalizeLegacyLocalQuery(value: String): String {
    if ('?' in value) return value
    val separator = value.indexOf('&')
    return if (separator < 0) value else value.substring(0, separator) + "?" + value.substring(separator + 1)
  }

  private fun parseCanonical(route: String, uri: Uri, original: String, requested: RequestedRuntime, source: RouteSource): LaunchDescriptor {
    if (!uri.host.equals("lynxview_page", true)) fail("unsupported_hybrid_host", "Unsupported Sparkling hybrid host '${uri.host.orEmpty()}'.")
    val items = queryItems(uri)
    val bundles = items.filter { it.name == "bundle" && !it.value.isNullOrEmpty() }
    val urls = items.filter { it.name == "url" && !it.value.isNullOrEmpty() }
    if (bundles.size + urls.size == 0) fail("missing_target", "The Sparkling route is missing bundle or url.")
    if (bundles.size + urls.size != 1) fail("ambiguous_target", "The Sparkling route must contain exactly one resource target.")
    val resource = if (bundles.isNotEmpty()) LaunchResource.LocalBundle(bundles.single().value!!) else {
      val value = urls.single().value!!
      val target = Uri.parse(value)
      if (target.scheme !in listOf("http", "https")) fail("malformed_sparkling_scheme", "Unsupported Sparkling URL target: $value")
      LaunchResource.RemoteBundle(value)
    }
    return descriptor(original, resource, requested, ResolvedRuntime.SPARKLING, source, route, items)
  }

  private fun descriptor(original: String, resource: LaunchResource, requested: RequestedRuntime, resolved: ResolvedRuntime, source: RouteSource, canonical: String?, items: List<LaunchQueryItem>): LaunchDescriptor {
    fun last(name: String) = items.lastOrNull { it.name == name && it.value != null }?.value
    fun bool(name: String) = last(name)?.let(::legacyBoolean)
    val fullscreen = bool("fullscreen") ?: false
    val hidden = fullscreen || (bool("hide_nav_bar") ?: bool("hidden_nav") ?: false)
    val width = legacyInt(last("width")); val height = legacyInt(last("height"))
    val viewport = if (width != null && height != null && width > 0 && height > 0) ViewportOptions(width, height) else null
    val extras = linkedMapOf<String, String>(); val props = linkedMapOf<String, Any>()
    items.forEach { it.value?.let { value -> extras[it.name] = value; props[camelCase(it.name)] = value } }
    return LaunchDescriptor(
      original, resource, last("initial_page"), viewport,
      AppearanceOptions(last("container_bg_color"), fullscreen || bool("trans_status_bar") == true, fullscreen || bool("hide_status_bar") == true, last("force_theme_style")),
      NavigationOptions(hidden, fullscreen, last("title"), last("title_color"), last("nav_bar_color") ?: last("bar_color"), last("back_button_style"), last("orientation")?.takeIf { it == "portrait" || it == "landscape" }),
      bool("animated") ?: true, last("enable_napi_addon")?.lowercase() in setOf("1", "true", "yes"), items.toList(), extras.toMap(), props.toMap(), requested, resolved, source, canonical)
  }

  private fun wrapperTarget(uri: Uri): String {
    val raw = uri.encodedQuery.orEmpty()
    val marker = "url="
    val starts = raw.split('&').count { it.startsWith(marker) }
    val index = raw.indexOf(marker)
    if (index < 0) fail("missing_target", "The route is missing its url target.")
    if (starts > 1) fail("ambiguous_target", "The route contains more than one url target.")
    val tail = raw.substring(index + marker.length)
    if (tail.substringBefore('&').contains("://") && tail.contains('&')) return decode(tail)
    val target = uri.getQueryParameter("url")
    if (target.isNullOrEmpty()) fail("missing_target", "The route is missing its url target.")
    return target
  }

  private fun queryItems(uri: Uri): List<LaunchQueryItem> {
    val query = uri.encodedQuery ?: return emptyList()
    return query.split('&').map { component ->
      val separator = component.indexOf('=')
      if (separator < 0) LaunchQueryItem(decode(component), null)
      else LaunchQueryItem(decode(component.substring(0, separator)), decode(component.substring(separator + 1)))
    }
  }

  private fun RequestedRuntime.resolveRaw() = if (this == RequestedRuntime.SPARKLING) ResolvedRuntime.SPARKLING else ResolvedRuntime.LYNX
  private fun decode(value: String): String {
    val decoded = try {
      URLDecoder.decode(value.replace("+", "%2B"), StandardCharsets.UTF_8.name())
    } catch (_: IllegalArgumentException) {
      fail("invalid_percent_encoding", "Invalid percent escape in route.")
    }
    if ('\uFFFD' in decoded) fail("invalid_percent_encoding", "The route contains invalid UTF-8 percent encoding.")
    return decoded
  }
  private fun legacyBoolean(value: String): Boolean = value.toDoubleOrNull()?.let { it != 0.0 } ?: value.equals("true", true) || value.equals("yes", true)
  private fun legacyInt(value: String?): Int? {
    if (value == null) return null
    return Regex("^[+-]?\\d+").find(value)?.value?.toLongOrNull()
      ?.takeIf { it in Int.MIN_VALUE..Int.MAX_VALUE }?.toInt()
  }
  private fun camelCase(name: String): String = name.trimStart('_').split('_').filter { it.isNotEmpty() }.let { parts -> parts.firstOrNull().orEmpty() + parts.drop(1).joinToString("") { it.replaceFirstChar(Char::uppercase) } }
  private fun validatePercentEncoding(value: String) {
    var i = 0
    while (i < value.length) { if (value[i] == '%') { if (i + 2 >= value.length || !value[i + 1].isHex() || !value[i + 2].isHex()) fail("invalid_percent_encoding", "Invalid percent escape in route."); i += 3 } else i++ }
  }
  private fun Char.isHex() = this in '0'..'9' || this in 'a'..'f' || this in 'A'..'F'
  private fun fail(code: String, message: String): Nothing = throw RouteException(code, message)
}
